lambda-buffers-proto/compiler.proto

```diff
@@ -491,3 +489,107 @@ message CompilerInput {
   // Modules to compile.
   repeated Module modules = 1;
 }
+
+
+// Compiler Failure can be extended with other classes of errors.
+message CompilerFailure {
+  // Reason why the compilation failed.
+  oneof failure {
+    KindCheckErr kc_err = 1;
+  }
+
+}
+
+/* Inference Failure 
+ */
+
+message InferenceErr {
+
+  // Error referring to an unbound term. This usually means that the term was
+  // not defined.
+  message UnboundTermErr {
+    string msg = 1; 
+  }
+
+  // This should never happen.
+  message ImpossibleErr {
+    string msg = 1;
+  }
+
+  // Unification has failed - type is incorrectly defined.
+  message UnificationErr {
+    string msg = 1; 
+  }
+
+  // Infinite recursion due to a substitution replacing a term with a term
+  // containing itself. Should never type check.
+  message RecursiveSubstitutionErr {
+    string msg = 1;
+  }
```

Comment:
When does this happen?

Reply:
This happens when the inference creates a substitution where the substituted term contains the variable its being substituted with. I.e.:

Substitute `a` with `a -> a`. Any such type would be an incorrect type.